cpp/src/arrow/flight/types.h

```diff
@@ -409,6 +409,15 @@ struct ARROW_FLIGHT_EXPORT Location {
   /// \brief Get the scheme of this URI.
   std::string scheme() const;
 
+  /// \brief Get the path of this URI.
+  std::string path() const;
+
+  /// \brief Get the query parameters of this URI.
+  arrow::Result<std::vector<std::pair<std::string, std::string>>> QueryItems() const;
+
+  /// \brief Convert URI path and parameters to headers.
+  arrow::Result<std::vector<std::pair<std::string, std::string>>> AsHeaders() const;
```

Comment:
This is still very specific to Flight SQL (and the docstring isn't very clear on what the transformation is). Could it be implemented as a helper function in the `arrow::flight::sql` namespace?

Reply:
Circling back around on this one, given it's part of the Location state/semantics I'm really inclined to have it kept here.  Additionally if anyone wanted to use the consistent Location behaviour for a Flight non-SQL impl it's best not to have them "borrowing" static stuff from fSQL, reimplementing it, or some other hacky means of getting at something that's intrinsically part of the Location object.  (If that's still an issue perhaps we can discuss briefly as I'm also sending an e-mail OOB somewhat related to this.)

Reply:
I don't think it's intrinsically part of a location: it's a specific way to transform a URI that makes sense in specific contexts. (Particularly because the notion of 'URI path' == 'catalog' really _only_ makes sense in Flight SQL, and it locks out other interpretations of the URI that other real users would like to have! #34829)

At the very least: the docstring is inadequate and I think the method name could be more specific as well. I would still prefer this as a static helper in the arrow::flight or arrow::flight::sql namespace but I'm willing to let that go if it's clearly documented what is happening here.

Reply:
NB we're probably dropping this due to various ways it ends up being inconsistent across e.g. Flight-based JDBC/ODBC drivers that already expose a catalog selection mechanism (and we don't need more duplicitious ways of doing X), as well as the fact that it doesn't get natively handled by the Flight SQL client but instead would have to percolate back out to the client app to be passed back in via SetSessionOptions.

Reply:
Ok.

I think since we never quite standardized the header-based parameters in the first place (as I recall, it was just something the JDBC driver happens to do), if you find it's simpler not to worry about compatibility, then I think that's OK. We can standardize around cookies/session tokens and assume the server will track things properly.

Reply:
It would be good to update the mailing list at some point just to get some eyes on it. (I'll review this when I get a chance. Thanks for bearing with me.)